utils/src/main/java/org/apache/cloudstack/utils/process/ProcessRunner.java

```diff
@@ -0,0 +1,112 @@
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+package org.apache.cloudstack.utils.process;
+
+import com.cloud.utils.concurrency.NamedThreadFactory;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import org.apache.log4j.Logger;
+import org.joda.time.Duration;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class ProcessRunner {
+    public static final Logger LOG = Logger.getLogger(ProcessRunner.class);
+
+    private static final ExecutorService processExecutor = Executors.newCachedThreadPool(new NamedThreadFactory("ProcessRunner"));
+
+    private static String readStream(final InputStream inputStream) {
+        String text = null;
+        try {
+            final BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(inputStream));
+            String line;
+            while ((line = bufferedReader.readLine()) != null) {
+                if (Strings.isNullOrEmpty(text)) {
+                    text = line;
+                } else {
+                    text = text + "\n" + line;
+                }
+            }
+        } catch (IOException e) {
+            if (LOG.isTraceEnabled()) {
+                LOG.trace("ProcessRunner::readStream failed due to: " + e.getMessage());
+            }
+        }
+        return text;
+    }
+
+    public static ProcessResult executeCommands(final List<String> commands, final Duration timeOut) {
+        Preconditions.checkArgument(commands != null && timeOut != null);
+
+        int retVal = -2;
+        String stdOutput = null;
+        String stdError = null;
+
+        try {
+            final Process process = new ProcessBuilder().command(commands).start();
+            if (timeOut.getStandardSeconds() > 0) {
```

Comment:
@rhtyd I don't think it acceptable to allow a process without any time boundary.  The concern is that a series of stalled commands could consume the thread pool and starve out other clients.   However, as a convenience to users of the class, it also makes senses that we would require the specification of a timeout on each call.  

A solution to this issue would be as follows:

1. Define a ``private final static Duration DEFAULT_COMMAND_TIMEOUT``
1. Provide an ``executeCommands`` override implementation:
```
public static ProcessResult executeCommands(final List<String> commands) {
    return executeCommands(commands, DEFAULT_COMMAND_TIMEOUT);
}
```
1. Modify the preconditions of this method to require a timeout greater than zero and simplify it to remove handling of that case.

This approach not only ensures that all commands are (softly) time-bounded, but it also avoid the [flag argument anti-pattern](http://martinfowler.com/bliki/FlagArgument.html).